common/src/test/java/io/netty/util/NettyRuntimeTests.java

```diff
@@ -0,0 +1,200 @@
+        final Runnable secondTarget = getRunnable(holder, barrier, secondRefernce);
+        final Thread secondGet = new Thread(secondTarget);
+        secondGet.start();
+
+        // release the hounds
+        await(barrier);
+
+        // wait for the hounds
+        await(barrier);
+
+        firstGet.join();
+        secondGet.join();
+
+        assertNull(firstReference.get());
+        assertNull(secondRefernce.get());
+    }
+
+    private Runnable getRunnable(
+            final NettyRuntime.AvailableProcessorsHolder holder,
+            final CyclicBarrier barrier,
+            final AtomicReference<IllegalStateException> reference) {
+        return new Runnable() {
+            @Override
+            public void run() {
+                await(barrier);
+                try {
+                    holder.availableProcessors();
+                } catch (final IllegalStateException e) {
+                    reference.set(e);
+                }
+                await(barrier);
+            }
+        };
+    }
+
+    @Test
+    public void testRacingGetAndSet() throws InterruptedException {
+        final NettyRuntime.AvailableProcessorsHolder holder = new NettyRuntime.AvailableProcessorsHolder();
+        final CyclicBarrier barrier = new CyclicBarrier(3);
+        final Thread get = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                await(barrier);
+                holder.availableProcessors();
+                await(barrier);
+            }
+        });
+        get.start();
+
+        final AtomicReference<IllegalStateException> setException = new AtomicReference<IllegalStateException>();
+        final Thread set = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                await(barrier);
+                try {
+                    holder.setAvailableProcessors(2048);
+                } catch (final IllegalStateException e) {
+                    setException.set(e);
+                }
+                await(barrier);
+            }
+        });
+        set.start();
+
+        // release the hounds
+        await(barrier);
+
+        // wait for the hounds
+        await(barrier);
+
+        get.join();
+        set.join();
+
+        if (setException.get() == null) {
+            assertThat(holder.availableProcessors(), equalTo(2048));
+        } else {
+            assertNotNull(setException.get());
+        }
+    }
+
+    @Test
+    public void testGetWithSystemProperty() {
+        final String availableProcessorsSystemProperty = SystemPropertyUtil.get("io.netty.availableProcessors");
+        System.setProperty("io.netty.availableProcessors", "2048");
+        final NettyRuntime.AvailableProcessorsHolder holder = new NettyRuntime.AvailableProcessorsHolder();
+        assertThat(holder.availableProcessors(), equalTo(2048));
+        if (availableProcessorsSystemProperty != null) {
```

Comment:
consider doing the restore of state in a final block as this may impact other tests if a failure occurs.